File: backtracking/9663-n-queen/1.py
import sys

input = lambda: sys.stdin.readline().rstrip()
write = lambda x: sys.stdout.write(str(x) + "\n")

def isPossible(x,y):
    for Q_y in range(y):
        Q_x = graph[Q_y]
        if Q_x == x or abs(x- Q_x) == y- Q_y:
            return False

    return True

def backTracking(y):
    global ans
    if y == n:
        ans += 1
        return

    for x in range(n):
        if isPossible(x,y):
            graph[y] = x
            backTracking(y+1)
            graph[y] = 0

    return
# 출력
n = int(input()) # 15
graph = [0] * n  # Q의 위치
ans = 0
backTracking(0)
print(ans)

File: backtracking/9663-n-queen/1.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import kotlin.math.abs

val br = BufferedReader(InputStreamReader(System.`in`))
val bw = BufferedWriter(OutputStreamWriter(System.out))
val n = br.readLine().toInt()
val graph = IntArray(n)
var ans = 0
fun main() {
    searchQueen(0)
    bw.write("$ans\n")
    bw.flush()
    bw.close()
    br.close()

}

private fun searchQueen(y: Int) {
    if (y == n) {
        ans++
        return
    }

    for (x in 0 until n) {
        if (isPossible(x, y)) {
            graph[y] = x
            searchQueen(y + 1)
            graph[y] = 0
        }
    }
}

/**
 * 현재 좌표 (x,y)에 queen을 놓을 수 있는가??
 * */
private fun isPossible(x: Int, y: Int): Boolean {
    for (q_y in 0 until y) {
        val q_x = graph[q_y] // 이전 queen의 x좌표

        // 1) x축 좌표가 겹치는지 체크
        // 2) 대각선 방향으로 공격경로가 겹치는지 체크
        if (q_x == x || abs(x - q_x) == y - q_y)
            return false
    }
    return true
}